server/base/src/main/java/org/apache/accumulo/server/util/upgrade/PreUpgradeCheck.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.accumulo.core.fate.zookeeper.ZooUtil.NodeMissingPolicy;
+import org.apache.accumulo.core.lock.ServiceLockPaths.AddressSelector;
+import org.apache.accumulo.core.lock.ServiceLockPaths.ServiceLockPath;
+import org.apache.accumulo.core.zookeeper.ZooSession;
+import org.apache.accumulo.server.AccumuloDataVersion;
+import org.apache.accumulo.server.ServerContext;
+import org.apache.accumulo.server.cli.ServerUtilOpts;
+import org.apache.accumulo.start.spi.KeywordExecutable;
+import org.apache.zookeeper.KeeperException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PreUpgradeCheck implements KeywordExecutable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PreUpgradeCheck.class);
+
+  @Override
+  public String keyword() {
+    return "pre-upgrade";
+  }
+
+  @Override
+  public String description() {
+    return "Utility that prepares an instance to be upgraded to a new version."
+        + " This utility must be run before starting any servers.";
+  }
+
+  @Override
+  public void execute(String[] args) throws Exception {
+
+    ServerUtilOpts opts = new ServerUtilOpts();
+    opts.parseArgs(PreUpgradeCheck.class.getName(), args);
+
+    final ServerContext context = opts.getServerContext();
+
+    final int persistentVersion = AccumuloDataVersion.getCurrentVersion(context);
+    final int thisVersion = AccumuloDataVersion.get();
+    if (persistentVersion == thisVersion) {
+      throw new IllegalStateException("Running this utility is unnecessary, this instance"
+          + " has already been upgraded to version " + thisVersion);
+    }
+
+    final ZooSession zs = context.getZooSession();
+    final ZooReader zr = zs.asReader();
+    final String prepUpgradePath = Constants.ZPREPARE_FOR_UPGRADE;
+
+    if (!zr.exists(prepUpgradePath)) {
+      LOG.info("{} node not found in ZooKeeper, 'ZooZap -prepare-for-upgrade' was likely"
+          + " not run after shutting down instance for upgrade. Removing"
+          + " server locks and checking for fate transactions.", prepUpgradePath);
+
+      try {
+        // Adapted from UpgradeCoordinator.abortIfFateTransactions
+        if (!zr.getChildren(Constants.ZFATE).isEmpty()) {
+          throw new IllegalStateException("Cannot continue pre-upgrade checks"
+              + " because FATE transactions exist. You can start a tserver, but"
+              + " not the Manager, with the old version of Accumulo then use "
+              + " the shell to delete completed transactions and fail pending"
+              + " or in-progress transactions. Once all of the FATE transactions"
+              + " have been removed you can retry this operation.");
+        }
+      } catch (KeeperException | InterruptedException e) {
+        throw new IllegalStateException("Error checking for existing FATE transactions", e);
+      }
+      LOG.info("No FATE transactions found");
+
+      // Forcefully delete all server locks
```

Comment:
Is this trying to kill older servers processes?  If so will this new code pick up older server paths?

Reply:
The idea is that the `ZooZap` changes in this PR would be backported to `3.1` and possibly `2.1.4`. The user would run `ZooZap -prepare-for-upgrade` and ZooZap would make sure that no fate transactions exist, remove server locks in ZooKeeper for that version, and create the `ZPREPARE_FOR_UPGRADE` node in ZooKeeper.

In this version of Accumulo, AbstractServer will prevent any server process from starting if ZPREPARE_FOR_UPGRADE exists. This code might be redundant, but yes, it's deleting all the server paths to clean them up and to remove the locks.

Reply:
Made a comment in the new code about this.